sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -374,8 +373,9 @@ private[execution] final class LongToUnsafeRowMap(var mm: TaskMemoryManager, cap
     // do not support spilling
     val got = mm.acquireExecutionMemory(size, MemoryMode.ON_HEAP, this)
     if (got < size) {
-      mm.releaseExecutionMemory(got, MemoryMode.ON_HEAP, this)
-      throw new SparkException(s"Can't acquire $size bytes memory to build hash relation")
+      freeMemory(got)
```

Comment:
yeah, all these changes suggested by the comments here: https://github.com/apache/spark/commit/90c0a04506a4972b7a2ac2b7dda0c5f8509a6e2f#commitcomment-17040679